### PR TITLE
[Merged by Bors] - feat: define `NonUnital(Semi)NormedCommRing`

### PR DESCRIPTION
**Pull request by j-loreaux**
This adds commutative versions of the existing `NonUnital(Semi)NormedRing` classes. This is essential for talking about, for example, non-unital commutative C⋆-algebras.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
!bench

**Comment by j-loreaux** [comment]
I ran the benchmark just to be sure, but I don't consider the results damning.

File: Mathlib/Analysis/Normed/Field/Basic.lean [hunk 1]
```diff
@@ -102,6 +102,24 @@ instance (priority := 100) NormedRing.toNonUnitalNormedRing [β : NormedRing α]
...
+norm which satisfies the inequality `‖x y‖ ≤ ‖x‖ ‖y‖`. -/
+class NonUnitalNormedCommRing (α : Type*) extends NonUnitalNormedRing α where
+  /-- Multiplication is commutative. -/
+  mul_comm : ∀ x y : α, x * y = y * x
+
+-- see Note [lower instance priority]
+/-- A non-unital normed commutative ring is a non-unital seminormed commutative ring. -/
+instance (priority := 100) NonUnitalNormedCommRing.toNonUnitalSeminormedCommRing
+    [β : NonUnitalNormedCommRing α] : NonUnitalSeminormedCommRing α :=
+  { β with }
```
Comment (YaelDillies) [inline Mathlib/Analysis/Normed/Field/Basic.lean:121]: eg
```suggestion
    [β : NonUnitalNormedCommRing α] : NonUnitalSeminormedCommRing α where
  toNonUnitalSeminormedRing :=
    NonUnitalNormedCommRing.toNonUnitalNormedRing.toNonUnitalSeminormedRing
  mul_comm := mul_comm
```

**Review by YaelDillies — COMMENTED** [review]
Does the benchmark get much better if you eliminate the `{ β with }` in favor of direct parent definitions? If not, I am happy to merge as is.

File: Mathlib/Analysis/Normed/Field/Basic.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/Analysis/Normed/Field/Basic.lean:121]: Good point, I'll try it.

**Comment by j-loreaux** [comment]
I've reverted the test.

**Comment by YaelDillies** [comment]
Let's merge as is but ping @eric-wieser in case he understands why the test was not much better.

maintainer merge

File: Mathlib/Analysis/Normed/Field/Basic.lean [hunk 2]
```diff
@@ -102,6 +102,24 @@ instance (priority := 100) NormedRing.toNonUnitalNormedRing [β : NormedRing α]
...
 
+/-- A non-unital seminormed commutative ring is a non-unital commutative ring endowed with a
+seminorm which satisfies the inequality `‖x y‖ ≤ ‖x‖ ‖y‖`. -/
+class NonUnitalSeminormedCommRing (α : Type*) extends NonUnitalSeminormedRing α where
+  /-- Multiplication is commutative. -/
+  mul_comm : ∀ x y : α, x * y = y * x
+
+/-- A non-unital normed commutative ring is a non-unital commutative ring endowed with a
+norm which satisfies the inequality `‖x y‖ ≤ ‖x‖ ‖y‖`. -/
+class NonUnitalNormedCommRing (α : Type*) extends NonUnitalNormedRing α where
```
Comment (eric-wieser) [inline Mathlib/Analysis/Normed/Field/Basic.lean:113]: Is there a reason not to write
```suggestion
class NonUnitalNormedCommRing (α : Type*) extends NonUnitalNormedRing α, NonUnitalSeminormedCommRing α where
```
(other than the line length limit!)

**Review by eric-wieser — APPROVED** [review]
bors merge

I think the `with` version is better from a maintenance point of view anyway, and we should be fixing the performance issues upstream or with flat structures.

**Pull request closed without merging**